File: templates/examples/core-components/redis/example-pub-sub.py
#!/usr/bin/env python3
"""
Redis Pub/Sub Example (Python)

This example demonstrates Redis publish/subscribe functionality
for real-time messaging and event-driven architectures.
"""

import os
import sys
import json
import time
import threading
from typing import Any, Dict, List, Optional, Callable
from dataclasses import dataclass, asdict
from datetime import datetime
import redis
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

@dataclass
class Message:
    """Message data model."""
    id: str
    channel: str
    content: str
    timestamp: str
    sender: str
    message_type: str = "text"
    metadata: Optional[Dict[str, Any]] = None

class RedisPubSub:
    """Redis publish/subscribe client."""
    
    def __init__(self, 
                 host: str = None,
                 port: int = None,
                 db: int = None,
                 password: str = None):
        """Initialize Redis pub/sub client."""
        
        # Get configuration from environment variables
        self.host = host or os.getenv("REDIS_HOST", "localhost")
        self.port = port or int(os.getenv("REDIS_PORT", "6379"))
        self.db = db or int(os.getenv("REDIS_DB", "0"))
        self.password = password or os.getenv("REDIS_PASSWORD")
        
        # Redis clients (separate for pub and sub)
        self.publisher = redis.Redis(
            host=self.host,
            port=self.port,
            db=self.db,
            password=self.password,
            decode_responses=True
        )
        
        self.subscriber = redis.Redis(
            host=self.host,
            port=self.port,
            db=self.db,
            password=self.password,
            decode_responses=True
        )
        
        # Test connection
        try:
            self.publisher.ping()
            print(f"✅ Redis Pub/Sub connected at {self.host}:{self.port}")
        except redis.ConnectionError as e:
            print(f"❌ Failed to connect to Redis: {e}")
            raise
        
        # Pub/Sub object
        self.pubsub = self.subscriber.pubsub()
        
        # Message handlers
        self.message_handlers: Dict[str, List[Callable]] = {}
        self.is_listening = False
        self.listener_thread = None
    
    def publish(self, channel: str, message: Any) -> int:
        """Publish a message to a channel."""
        try:
            # Serialize message if it's not a string
            if not isinstance(message, str):
                message = json.dumps(message)
            
            # Publish message
            subscriber_count = self.publisher.publish(channel, message)
            
            print(f"📡 Published to {channel}: {subscriber_count} subscribers")
            return subscriber_count
            
        except Exception as e:
            print(f"❌ Error publishing to {channel}: {e}")
            return 0
    
    def subscribe(self, *channels: str) -> bool:
        """Subscribe to one or more channels."""
        try:
            self.pubsub.subscribe(*channels)
            print(f"🔔 Subscribed to channels: {', '.join(channels)}")
            return True
            
        except Exception as e:
            print(f"❌ Error subscribing to channels: {e}")
            return False
    
    def unsubscribe(self, *channels: str) -> bool:
        """Unsubscribe from one or more channels."""
        try:
            self.pubsub.unsubscribe(*channels)
            print(f"🔕 Unsubscribed from channels: {', '.join(channels)}")
            return True
            
        except Exception as e:
            print(f"❌ Error unsubscribing from channels: {e}")
            return False
    
    def pattern_subscribe(self, *patterns: str) -> bool:
        """Subscribe to channel patterns."""
        try:
            self.pubsub.psubscribe(*patterns)
            print(f"🔔 Subscribed to patterns: {', '.join(patterns)}")
            return True
            
        except Exception as e:
            print(f"❌ Error subscribing to patterns: {e}")
            return False
    
    def add_message_handler(self, channel: str, handler: Callable[[str, Any], None]):
        """Add a message handler for a specific channel."""
        if channel not in self.message_handlers:
            self.message_handlers[channel] = []
        
        self.message_handlers[channel].append(handler)
        print(f"🔧 Added message handler for {channel}")
    
    def remove_message_handler(self, channel: str, handler: Callable[[str, Any], None]):
        """Remove a message handler for a specific channel."""
        if channel in self.message_handlers:
            try:
                self.message_handlers[channel].remove(handler)
                print(f"🔧 Removed message handler for {channel}")
            except ValueError:
                print(f"⚠️ Handler not found for {channel}")
    
    def start_listening(self):
        """Start listening for messages in a separate thread."""
        if self.is_listening:
            print("⚠️ Already listening for messages")
            return
        
        self.is_listening = True
        self.listener_thread = threading.Thread(target=self._listen_for_messages, daemon=True)
        self.listener_thread.start()
        print("🎧 Started listening for messages")
    
    def stop_listening(self):
        """Stop listening for messages."""
        if not self.is_listening:
            print("⚠️ Not currently listening")
            return
        
        self.is_listening = False
        if self.listener_thread:
            self.listener_thread.join(timeout=1)
        
        print("🛑 Stopped listening for messages")
    
    def _listen_for_messages(self):
        """Listen for messages (runs in separate thread)."""
        try:
            for message in self.pubsub.listen():
                if not self.is_listening:
                    break
                
                # Handle different message types
                if message['type'] == 'message':
                    self._handle_message(message['channel'], message['data'])
                elif message['type'] == 'pmessage':
                    self._handle_pattern_message(message['pattern'], message['channel'], message['data'])
                elif message['type'] == 'subscribe':
                    print(f"✅ Subscribed to {message['channel']}")
                elif message['type'] == 'unsubscribe':
                    print(f"❌ Unsubscribed from {message['channel']}")
                elif message['type'] == 'psubscribe':
                    print(f"✅ Pattern subscribed to {message['pattern']}")
                elif message['type'] == 'punsubscribe':
                    print(f"❌ Pattern unsubscribed from {message['pattern']}")
        
        except Exception as e:
            print(f"❌ Error in message listener: {e}")
    
    def _handle_message(self, channel: str, data: str):
        """Handle a regular message."""
        try:
            # Try to parse as JSON
            try:
                message_data = json.loads(data)
            except json.JSONDecodeError:
                message_data = data
            
            # Call handlers for this channel
            if channel in self.message_handlers:
                for handler in self.message_handlers[channel]:
                    try:
                        handler(channel, message_data)
                    except Exception as e:
                        print(f"❌ Error in message handler for {channel}: {e}")
            
            # Default handler
            print(f"📨 {channel}: {message_data}")
            
        except Exception as e:
            print(f"❌ Error handling message: {e}")
    
    def _handle_pattern_message(self, pattern: str, channel: str, data: str):
        """Handle a pattern message."""
        try:
            # Try to parse as JSON
            try:
                message_data = json.loads(data)
            except json.JSONDecodeError:
                message_data = data
            
            print(f"📨 Pattern {pattern} -> {channel}: {message_data}")
            
        except Exception as e:
            print(f"❌ Error handling pattern message: {e}")
    
    def get_subscribers_count(self, channel: str) -> int:
        """Get number of subscribers for a channel."""
        try:
            result = self.publisher.pubsub_numsub(channel)
            return result[0][1] if result else 0
        except Exception as e:
            print(f"❌ Error getting subscriber count: {e}")
            return 0
    
    def get_active_channels(self) -> List[str]:
        """Get list of active channels."""
        try:
            return self.publisher.pubsub_channels()
        except Exception as e:
            print(f"❌ Error getting active channels: {e}")
            return []

class ChatRoom:
    """Chat room implementation using Redis pub/sub."""
    
    def __init__(self, room_name: str):
        """Initialize chat room."""
        self.room_name = room_name
        self.channel = f"chat:{room_name}"
        self.pubsub = RedisPubSub()
        self.users: Dict[str, datetime] = {}
        
        # Add message handler
        self.pubsub.add_message_handler(self.channel, self._handle_chat_message)
        
        # Subscribe to chat channel
        self.pubsub.subscribe(self.channel)
    
    def join(self, username: str):
        """Join the chat room."""
        self.users[username] = datetime.now()
        
        # Send join message
        join_message = Message(
            id=f"join_{int(time.time())}",
            channel=self.channel,
            content=f"{username} joined the room",
            timestamp=datetime.now().isoformat(),
            sender="system",
            message_type="join"
        )
        
        self.pubsub.publish(self.channel, asdict(join_message))
        print(f"👋 {username} joined {self.room_name}")
    
    def leave(self, username: str):
        """Leave the chat room."""
        if username in self.users:
            del self.users[username]
            
            # Send leave message
            leave_message = Message(
                id=f"leave_{int(time.time())}",
                channel=self.channel,
                content=f"{username} left the room",
                timestamp=datetime.now().isoformat(),
                sender="system",
                message_type="leave"
            )
            
            self.pubsub.publish(self.channel, asdict(leave_message))
            print(f"👋 {username} left {self.room_name}")
    
    def send_message(self, sender: str, content: str):
        """Send a chat message."""
        message = Message(
            id=f"msg_{int(time.time())}",
            channel=self.channel,
            content=content,
            timestamp=datetime.now().isoformat(),
            sender=sender,
            message_type="chat"
        )
        
        self.pubsub.publish(self.channel, asdict(message))
    
    def _handle_chat_message(self, channel: str, data: Dict[str, Any]):
        """Handle chat messages."""
        try:
            message = Message(**data)
            
            if message.message_type == "chat":
                print(f"💬 [{message.sender}]: {message.content}")
            elif message.message_type == "join":
                print(f"✅ {message.content}")
            elif message.message_type == "leave":
                print(f"❌ {message.content}")
            
        except Exception as e:
            print(f"❌ Error handling chat message: {e}")
    
    def start_listening(self):
        """Start listening for chat messages."""
        self.pubsub.start_listening()
    
    def stop_listening(self):
        """Stop listening for chat messages."""
        self.pubsub.stop_listening()

class EventBus:
    """Event bus implementation using Redis pub/sub."""
    
    def __init__(self):
        """Initialize event bus."""
        self.pubsub = RedisPubSub()
        self.event_handlers: Dict[str, List[Callable]] = {}
    
    def subscribe_to_event(self, event_type: str, handler: Callable[[Dict[str, Any]], None]):
        """Subscribe to an event type."""
        channel = f"events:{event_type}"
        
        if event_type not in self.event_handlers:
            self.event_handlers[event_type] = []
            self.pubsub.subscribe(channel)
        
        self.event_handlers[event_type].append(handler)
        self.pubsub.add_message_handler(channel, lambda ch, data: self._handle_event(event_type, data))
    
    def publish_event(self, event_type: str, event_data: Dict[str, Any]):
        """Publish an event."""
        channel = f"events:{event_type}"
        
        event = {
            "type": event_type,
            "data": event_data,
            "timestamp": datetime.now().isoformat(),
            "id": f"event_{int(time.time())}"
        }
        
        self.pubsub.publish(channel, event)
    
    def _handle_event(self, event_type: str, data: Dict[str, Any]):
        """Handle events."""
        if event_type in self.event_handlers:
            for handler in self.event_handlers[event_type]:
                try:
                    handler(data)
                except Exception as e:
                    print(f"❌ Error in event handler for {event_type}: {e}")
    
    def start_listening(self):
        """Start listening for events."""
        self.pubsub.start_listening()
    
    def stop_listening(self):
        """Stop listening for events."""
        self.pubsub.stop_listening()

def main():
    """Main function demonstrating Redis pub/sub."""
    print("📡 Redis Pub/Sub Example (Python)")
    print("=" * 50)
    
    try:
        # Initialize pub/sub
        pubsub = RedisPubSub()
        
        # Demo 1: Basic pub/sub
        print("\n🔔 Basic Pub/Sub Demo")
        print("-" * 30)
        
        # Subscribe to channels
        pubsub.subscribe("news", "alerts")
        pubsub.start_listening()
        
        # Give it a moment to start
        time.sleep(1)
        
        # Publish some messages
        pubsub.publish("news", "Breaking: Redis pub/sub working!")
        pubsub.publish("alerts", "System maintenance in 1 hour")
        pubsub.publish("news", {"headline": "JSON message test", "priority": "high"})
        
        time.sleep(2)
        
        # Demo 2: Chat room
        print("\n💬 Chat Room Demo")
        print("-" * 30)
        
        chat_room = ChatRoom("general")
        chat_room.start_listening()
        
        # Simulate chat activity
        chat_room.join("alice")
        time.sleep(1)
        chat_room.join("bob")
        time.sleep(1)
        
        chat_room.send_message("alice", "Hello everyone!")
        time.sleep(1)
        chat_room.send_message("bob", "Hi Alice! How are you?")
        time.sleep(1)
        chat_room.send_message("alice", "I'm doing great, thanks!")
        
        time.sleep(2)
        
        # Demo 3: Event bus
        print("\n🚌 Event Bus Demo")
        print("-" * 30)
        
        event_bus = EventBus()
        
        # Subscribe to events
        def handle_user_signup(event_data):
            print(f"🆕 New user signup: {event_data['data']['username']}")
        
        def handle_order_placed(event_data):
            print(f"🛒 Order placed: ${event_data['data']['amount']} by {event_data['data']['customer']}")
        
        event_bus.subscribe_to_event("user_signup", handle_user_signup)
        event_bus.subscribe_to_event("order_placed", handle_order_placed)
        event_bus.start_listening()
        
        time.sleep(1)
        
        # Publish events
        event_bus.publish_event("user_signup", {"username": "john_doe", "email": "john@example.com"})
        event_bus.publish_event("order_placed", {"customer": "alice", "amount": 49.99, "items": 3})
        
        time.sleep(2)
        
        # Interactive mode
        print("\n" + "=" * 50)
        print("🎯 Interactive Pub/Sub Demo")
        print("Commands:")
        print("  'sub <channel>' - Subscribe to channel")
        print("  'unsub <channel>' - Unsubscribe from channel")
        print("  'pub <channel> <message>' - Publish message")
        print("  'psub <pattern>' - Subscribe to pattern")
        print("  'count <channel>' - Get subscriber count")
        print("  'channels' - List active channels")
        print("  'quit' - Exit")
        print("-" * 50)
        
        while True:
            user_input = input("\n📡 PubSub> ").strip()
            
            if user_input.lower() == 'quit':
                print("👋 Goodbye!")
                break
            elif user_input.startswith('sub '):
                channel = user_input.split(' ', 1)[1]
                pubsub.subscribe(channel)
            elif user_input.startswith('unsub '):
                channel = user_input.split(' ', 1)[1]
                pubsub.unsubscribe(channel)
            elif user_input.startswith('pub '):
                parts = user_input.split(' ', 2)
                if len(parts) >= 3:
                    channel, message = parts[1], parts[2]
                    pubsub.publish(channel, message)
                else:
                    print("Usage: pub <channel> <message>")
            elif user_input.startswith('psub '):
                pattern = user_input.split(' ', 1)[1]
                pubsub.pattern_subscribe(pattern)
            elif user_input.startswith('count '):
                channel = user_input.split(' ', 1)[1]
                count = pubsub.get_subscribers_count(channel)
                print(f"👥 {channel}: {count} subscribers")
            elif user_input.lower() == 'channels':
                channels = pubsub.get_active_channels()
                print(f"📺 Active channels: {', '.join(channels) if channels else 'None'}")
            elif user_input:
                print("Unknown command. Type 'quit' to exit.")
        
        # Cleanup
        pubsub.stop_listening()
        chat_room.stop_listening()
        event_bus.stop_listening()
        
    except KeyboardInterrupt:
        print("\n\n👋 Demo interrupted by user")
    except Exception as e:
        print(f"\n❌ Unexpected error: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: templates/examples/core-components/redis/example-pub-sub.js
/**
 * Redis Pub/Sub Examples
 * 
 * This example demonstrates:
 * - Real-time messaging with Redis pub/sub
 * - Channel subscriptions and pattern matching
 * - Message broadcasting
 * - Event-driven architecture
 * - WebSocket integration patterns
 * 
 * Usage:
 * npm install redis
 * node example-pub-sub.js
 */

const redis = require('redis');
const { EventEmitter } = require('events');

// Create separate Redis clients for pub/sub
const publisherClient = redis.createClient({
    host: process.env.REDIS_HOST || 'localhost',
    port: process.env.REDIS_PORT || 6379
});

const subscriberClient = redis.createClient({
    host: process.env.REDIS_HOST || 'localhost',
    port: process.env.REDIS_PORT || 6379
});

// Error handling
publisherClient.on('error', (err) => console.error('Publisher Error:', err));
subscriberClient.on('error', (err) => console.error('Subscriber Error:', err));

// ==========================================
// MESSAGE PUBLISHER
// ==========================================

class MessagePublisher {
    constructor(redisClient) {
        this.client = redisClient;
    }

    /**
     * Publish message to a channel
     */
    async publish(channel, message) {
        const messageData = {
            id: Date.now().toString(),
            timestamp: new Date().toISOString(),
            channel,
            payload: message
        };

        const serialized = JSON.stringify(messageData);
        const subscriberCount = await this.client.publish(channel, serialized);
        
        console.log(`📤 Published to '${channel}': ${JSON.stringify(message)} (${subscriberCount} subscribers)`);
        return subscriberCount;
    }

    /**
     * Broadcast message to multiple channels
     */
    async broadcast(channels, message) {
        const promises = channels.map(channel => this.publish(channel, message));
        const results = await Promise.all(promises);
        
        const totalSubscribers = results.reduce((sum, count) => sum + count, 0);
        console.log(`📡 Broadcasted to ${channels.length} channels, reached ${totalSubscribers} subscribers`);
        
        return totalSubscribers;
    }

    /**
     * Publish user-specific message
     */
    async publishToUser(userId, message) {
        const channel = `user:${userId}`;
        return await this.publish(channel, message);
    }

    /**
     * Publish room-specific message
     */
    async publishToRoom(roomId, message) {
        const channel = `room:${roomId}`;
        return await this.publish(channel, message);
    }

    /**
     * Publish system-wide notification
     */
    async publishSystemMessage(message) {
        const channel = 'system:notifications';
        return await this.publish(channel, message);
    }
}

// ==========================================
// MESSAGE SUBSCRIBER
// ==========================================

class MessageSubscriber extends EventEmitter {
    constructor(redisClient) {
        super();
        this.client = redisClient;
        this.subscriptions = new Set();
        this.patternSubscriptions = new Set();
        
        // Handle incoming messages
        this.client.on('message', (channel, message) => {
            this.handleMessage(channel, message);
        });
        
        // Handle pattern messages
        this.client.on('pmessage', (pattern, channel, message) => {
            this.handlePatternMessage(pattern, channel, message);
        });
    }

    /**
     * Subscribe to a channel
     */
    async subscribe(channel) {
        if (!this.subscriptions.has(channel)) {
            await this.client.subscribe(channel);
            this.subscriptions.add(channel);
            console.log(`📥 Subscribed to channel: ${channel}`);
        }
    }

    /**
     * Subscribe to multiple channels
     */
    async subscribeToChannels(channels) {
        const newChannels = channels.filter(channel => !this.subscriptions.has(channel));
        
        if (newChannels.length > 0) {
            await this.client.subscribe(...newChannels);
            newChannels.forEach(channel => this.subscriptions.add(channel));
            console.log(`📥 Subscribed to ${newChannels.length} channels:`, newChannels);
        }
    }

    /**
     * Subscribe to pattern
     */
    async subscribeToPattern(pattern) {
        if (!this.patternSubscriptions.has(pattern)) {
            await this.client.psubscribe(pattern);
            this.patternSubscriptions.add(pattern);
            console.log(`🔍 Subscribed to pattern: ${pattern}`);
        }
    }

    /**
     * Unsubscribe from channel
     */
    async unsubscribe(channel) {
        if (this.subscriptions.has(channel)) {
            await this.client.unsubscribe(channel);
            this.subscriptions.delete(channel);
            console.log(`❌ Unsubscribed from channel: ${channel}`);
        }
    }

    /**
     * Unsubscribe from pattern
     */
    async unsubscribeFromPattern(pattern) {
        if (this.patternSubscriptions.has(pattern)) {
            await this.client.punsubscribe(pattern);
            this.patternSubscriptions.delete(pattern);
            console.log(`❌ Unsubscribed from pattern: ${pattern}`);
        }
    }

    /**
     * Handle incoming message
     */
    handleMessage(channel, message) {
        try {
            const parsedMessage = JSON.parse(message);
            console.log(`📨 Received on '${channel}':`, parsedMessage.payload);
            
            // Emit events for different channel types
            this.emit('message', channel, parsedMessage);
            this.emit(`channel:${channel}`, parsedMessage);
            
            // Emit specific events based on channel naming
            if (channel.startsWith('user:')) {
                const userId = channel.split(':')[1];
                this.emit('userMessage', userId, parsedMessage);
            } else if (channel.startsWith('room:')) {
                const roomId = channel.split(':')[1];
                this.emit('roomMessage', roomId, parsedMessage);
            } else if (channel.startsWith('system:')) {
                this.emit('systemMessage', parsedMessage);
            }
            
        } catch (error) {
            console.error('Error parsing message:', error);
        }
    }

    /**
     * Handle pattern message
     */
    handlePatternMessage(pattern, channel, message) {
        try {
            const parsedMessage = JSON.parse(message);
            console.log(`🔍 Pattern '${pattern}' matched '${channel}':`, parsedMessage.payload);
            
            this.emit('patternMessage', pattern, channel, parsedMessage);
            this.emit(`pattern:${pattern}`, channel, parsedMessage);
            
        } catch (error) {
            console.error('Error parsing pattern message:', error);
        }
    }

    /**
     * Get active subscriptions
     */
    getSubscriptions() {
        return {
            channels: Array.from(this.subscriptions),
            patterns: Array.from(this.patternSubscriptions)
        };
    }
}

// ==========================================
// REAL-TIME CHAT SYSTEM
// ==========================================

class ChatSystem {
    constructor(publisher, subscriber) {
        this.publisher = publisher;
        this.subscriber = subscriber;
        this.rooms = new Map();
        this.users = new Map();
        
        this.setupEventHandlers();
    }

    setupEventHandlers() {
        // Handle room messages
        this.subscriber.on('roomMessage', (roomId, message) => {
            this.handleRoomMessage(roomId, message);
        });

        // Handle user messages
        this.subscriber.on('userMessage', (userId, message) => {
            this.handleUserMessage(userId, message);
        });

        // Handle system messages
        this.subscriber.on('systemMessage', (message) => {
            this.handleSystemMessage(message);
        });
    }

    /**
     * User joins a room
     */
    async joinRoom(userId, roomId, userData = {}) {
        const roomChannel = `room:${roomId}`;
        
        // Subscribe to room
        await this.subscriber.subscribe(roomChannel);
        
        // Track user in room
        if (!this.rooms.has(roomId)) {
            this.rooms.set(roomId, new Set());
        }
        this.rooms.get(roomId).add(userId);
        
        // Track user data
        this.users.set(userId, { ...userData, currentRoom: roomId });
        
        // Notify room about new user
        await this.publisher.publishToRoom(roomId, {
            type: 'user_joined',
            userId,
            userData,
            timestamp: new Date().toISOString()
        });
        
        console.log(`👤 User ${userId} joined room ${roomId}`);
    }

    /**
     * User leaves a room
     */
    async leaveRoom(userId, roomId) {
        const roomChannel = `room:${roomId}`;
        
        // Remove user from room tracking
        if (this.rooms.has(roomId)) {
            this.rooms.get(roomId).delete(userId);
            
            // Remove empty rooms
            if (this.rooms.get(roomId).size === 0) {
                this.rooms.delete(roomId);
                await this.subscriber.unsubscribe(roomChannel);
            }
        }
        
        // Update user data
        const userData = this.users.get(userId);
        if (userData) {
            delete userData.currentRoom;
        }
        
        // Notify room about user leaving
        await this.publisher.publishToRoom(roomId, {
            type: 'user_left',
            userId,
            timestamp: new Date().toISOString()
        });
        
        console.log(`👤 User ${userId} left room ${roomId}`);
    }

    /**
     * Send message to room
     */
    async sendMessageToRoom(userId, roomId, message) {
        const userData = this.users.get(userId);
        
        await this.publisher.publishToRoom(roomId, {
            type: 'chat_message',
            userId,
            username: userData?.username || `User${userId}`,
            message,
            timestamp: new Date().toISOString()
        });
    }

    /**
     * Send private message to user
     */
    async sendPrivateMessage(fromUserId, toUserId, message) {
        const fromUser = this.users.get(fromUserId);
        
        await this.publisher.publishToUser(toUserId, {
            type: 'private_message',
            fromUserId,
            fromUsername: fromUser?.username || `User${fromUserId}`,
            message,
            timestamp: new Date().toISOString()
        });
    }

    /**
     * Handle room message
     */
    handleRoomMessage(roomId, message) {
        const { type, userId, username, message: content } = message.payload;
        
        switch (type) {
            case 'user_joined':
                console.log(`🟢 ${username || userId} joined room ${roomId}`);
                break;
            case 'user_left':
                console.log(`🔴 ${username || userId} left room ${roomId}`);
                break;
            case 'chat_message':
                console.log(`💬 [${roomId}] ${username}: ${content}`);
                break;
        }
    }

    /**
     * Handle user message
     */
    handleUserMessage(userId, message) {
        const { type, fromUsername, message: content } = message.payload;
        
        if (type === 'private_message') {
            console.log(`📩 Private message to ${userId} from ${fromUsername}: ${content}`);
        }
    }

    /**
     * Handle system message
     */
    handleSystemMessage(message) {
        console.log(`🔔 System: ${message.payload.message}`);
    }

    /**
     * Get room statistics
     */
    getRoomStats() {
        const stats = {};
        for (const [roomId, users] of this.rooms) {
            stats[roomId] = {
                userCount: users.size,
                users: Array.from(users)
            };
        }
        return stats;
    }
}

// ==========================================
// NOTIFICATION SYSTEM
// ==========================================

class NotificationSystem {
    constructor(publisher, subscriber) {
        this.publisher = publisher;
        this.subscriber = subscriber;
        this.userSubscriptions = new Map();
        
        this.setupSubscriptions();
    }

    async setupSubscriptions() {
        // Subscribe to all notification patterns
        await this.subscriber.subscribeToPattern('notification:*');
        await this.subscriber.subscribeToPattern('user:*:notification');
        
        this.subscriber.on('patternMessage', (pattern, channel, message) => {
            this.handleNotification(pattern, channel, message);
        });
    }

    /**
     * Subscribe user to notifications
     */
    async subscribeUser(userId, notificationTypes = ['all']) {
        const userChannel = `user:${userId}:notification`;
        await this.subscriber.subscribe(userChannel);
        
        this.userSubscriptions.set(userId, notificationTypes);
        console.log(`🔔 User ${userId} subscribed to notifications:`, notificationTypes);
    }

    /**
     * Send notification to user
     */
    async sendUserNotification(userId, notification) {
        const channel = `user:${userId}:notification`;
        await this.publisher.publish(channel, {
            type: 'user_notification',
            userId,
            notification,
            timestamp: new Date().toISOString()
        });
    }

    /**
     * Send notification to all users
     */
    async sendGlobalNotification(notification) {
        const channel = 'notification:global';
        await this.publisher.publish(channel, {
            type: 'global_notification',
            notification,
            timestamp: new Date().toISOString()
        });
    }

    /**
     * Send notification by type
     */
    async sendNotificationByType(type, notification) {
        const channel = `notification:${type}`;
        await this.publisher.publish(channel, {
            type: 'typed_notification',
            notificationType: type,
            notification,
            timestamp: new Date().toISOString()
        });
    }

    /**
     * Handle notification
     */
    handleNotification(pattern, channel, message) {
        const { type, notification } = message.payload;
        
        switch (type) {
            case 'user_notification':
                console.log(`🔔 User notification: ${notification.title}`);
                break;
            case 'global_notification':
                console.log(`📢 Global notification: ${notification.title}`);
                break;
            case 'typed_notification':
                console.log(`🏷️ ${message.payload.notificationType} notification: ${notification.title}`);
                break;
        }
    }
}

// ==========================================
// REAL-TIME ANALYTICS
// ==========================================

class AnalyticsSystem {
    constructor(publisher, subscriber) {
        this.publisher = publisher;
        this.subscriber = subscriber;
        this.metrics = new Map();
        
        this.setupAnalytics();
    }

    async setupAnalytics() {
        // Subscribe to analytics events
        await this.subscriber.subscribeToPattern('analytics:*');
        
        this.subscriber.on('patternMessage', (pattern, channel, message) => {
            if (pattern === 'analytics:*') {
                this.handleAnalyticsEvent(channel, message);
            }
        });
    }

    /**
     * Track event
     */
    async trackEvent(eventType, data) {
        const channel = `analytics:${eventType}`;
        await this.publisher.publish(channel, {
            eventType,
            data,
            timestamp: new Date().toISOString()
        });
    }

    /**
     * Track user action
     */
    async trackUserAction(userId, action, metadata = {}) {
        await this.trackEvent('user_action', {
            userId,
            action,
            metadata
        });
    }

    /**
     * Track page view
     */
    async trackPageView(userId, page, metadata = {}) {
        await this.trackEvent('page_view', {
            userId,
            page,
            metadata
        });
    }

    /**
     * Handle analytics event
     */
    handleAnalyticsEvent(channel, message) {
        const { eventType, data } = message.payload;
        
        // Update metrics
        if (!this.metrics.has(eventType)) {
            this.metrics.set(eventType, { count: 0, lastSeen: null });
        }
        
        const metric = this.metrics.get(eventType);
        metric.count++;
        metric.lastSeen = new Date().toISOString();
        
        console.log(`📊 Analytics: ${eventType} (${metric.count} times)`);
    }

    /**
     * Get analytics summary
     */
    getAnalyticsSummary() {
        const summary = {};
        for (const [eventType, metric] of this.metrics) {
            summary[eventType] = metric;
        }
        return summary;
    }
}

// ==========================================
// EXAMPLE USAGE
// ==========================================

async function runPubSubExamples() {
    try {
        // Initialize components
        const publisher = new MessagePublisher(publisherClient);
        const subscriber = new MessageSubscriber(subscriberClient);
        
        console.log('=== Basic Pub/Sub Example ===');
        
        // Subscribe to channels
        await subscriber.subscribe('news');
        await subscriber.subscribe('alerts');
        
        // Set up message handlers
        subscriber.on('channel:news', (message) => {
            console.log('📰 News received:', message.payload);
        });
        
        subscriber.on('channel:alerts', (message) => {
            console.log('🚨 Alert received:', message.payload);
        });
        
        // Publish messages
        await publisher.publish('news', { title: 'Breaking News', content: 'Something important happened' });
        await publisher.publish('alerts', { type: 'warning', message: 'System maintenance scheduled' });
        
        // Wait for messages to be processed
        await new Promise(resolve => setTimeout(resolve, 100));

        console.log('\\n=== Pattern Subscription Example ===');
        
        // Subscribe to patterns
        await subscriber.subscribeToPattern('user:*');
        await subscriber.subscribeToPattern('room:*');
        
        // Publish to matching channels
        await publisher.publish('user:123', { type: 'profile_update', data: { name: 'John' } });
        await publisher.publish('room:general', { type: 'message', text: 'Hello everyone!' });
        
        await new Promise(resolve => setTimeout(resolve, 100));

        console.log('\\n=== Chat System Example ===');
        
        // Create chat system
        const chatSystem = new ChatSystem(publisher, subscriber);
        
        // Users join rooms
        await chatSystem.joinRoom('user1', 'general', { username: 'Alice' });
        await chatSystem.joinRoom('user2', 'general', { username: 'Bob' });
        
        // Send messages
        await chatSystem.sendMessageToRoom('user1', 'general', 'Hello everyone!');
        await chatSystem.sendMessageToRoom('user2', 'general', 'Hi Alice!');
        
        // Send private message
        await chatSystem.sendPrivateMessage('user1', 'user2', 'Hey Bob, how are you?');
        
        await new Promise(resolve => setTimeout(resolve, 100));

        console.log('\\n=== Notification System Example ===');
        
        // Create notification system
        const notificationSystem = new NotificationSystem(publisher, subscriber);
        
        // Subscribe user to notifications
        await notificationSystem.subscribeUser('user1', ['security', 'updates']);
        
        // Send notifications
        await notificationSystem.sendUserNotification('user1', {
            title: 'Security Alert',
            body: 'New login detected'
        });
        
        await notificationSystem.sendGlobalNotification({
            title: 'System Update',
            body: 'System will be updated tonight'
        });
        
        await new Promise(resolve => setTimeout(resolve, 100));

        console.log('\\n=== Analytics System Example ===');
        
        // Create analytics system
        const analyticsSystem = new AnalyticsSystem(publisher, subscriber);
        
        // Track events
        await analyticsSystem.trackUserAction('user1', 'login', { ip: '192.168.1.100' });
        await analyticsSystem.trackPageView('user1', '/dashboard', { referrer: '/login' });
        await analyticsSystem.trackEvent('purchase', { userId: 'user1', amount: 99.99 });
        
        await new Promise(resolve => setTimeout(resolve, 100));
        
        console.log('\\n=== Final Statistics ===');
        console.log('Room Stats:', chatSystem.getRoomStats());
        console.log('Analytics Summary:', analyticsSystem.getAnalyticsSummary());
        console.log('Subscriptions:', subscriber.getSubscriptions());
        
    } catch (error) {
        console.error('Error in pub/sub examples:', error);
    }
}

// Run examples when script is executed directly
if (require.main === module) {
    Promise.all([
        new Promise(resolve => publisherClient.on('ready', resolve)),
        new Promise(resolve => subscriberClient.on('ready', resolve))
    ]).then(async () => {
        await runPubSubExamples();
        
        // Close connections
        publisherClient.quit();
        subscriberClient.quit();
    });
}

module.exports = {
    MessagePublisher,
    MessageSubscriber,
    ChatSystem,
    NotificationSystem,
    AnalyticsSystem
};